### PR TITLE
Embeds: Don't transform into specialized embed block variation if it's not registered

### PR DESCRIPTION
**Pull request by ockham**
## Description
Users can insert a generic version of the `core/embed` block from the block inserter, or choose a more specialized block variation (e.g. for Instagram, YouTube, etc).

Furthermore, if a user inserts a plain URL, we also trigger our embed block creation behavior: We first create a generic `core/embed` block from that URL, and then analyze if we have a more specialized block variation. If we do, we transform the generic block into that block variation.

A site admit might now choose to unregister certain block variations, in order to prevent users from inserting embeds of a certain type (e.g. they might choose to disallow WordPress embeds out of security concerns).

In that case however, the URL insertion behavior will still attempt to insert a embed block variation for that specific embed type, even though that block variation is not available.

We should thus check if a block variation of the `core/embed` is registered for the given embed type before attempting to transform the generic embed block into that variation.

## Related conversations

https://github.com/WordPress/gutenberg/pull/24090#issuecomment-673705516
https://github.com/Automattic/wp-calypso/issues/39935#issuecomment-671825119

## How has this been tested?

Verify that embeds still work.

The specific use case where a block variation isn't registered should be covered by the unit tests, but you can add some code to unregister a variation manually and insert a matching embed to verify that the generic embed block isn't tranformed to that variation.

## Types of changes
Bug fix

## Checklist:
- [ ] My code is tested.
- [ ] My code follows the WordPress code style. 
- [ ] My code follows the accessibility standards. 
- [ ] My code has proper inline documentation. 
- [ ] I've included developer documentation if appropriate. 
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR.

### CODE REVIEW

File: packages/block-library/src/embed/util.js [hunk 1]
```diff
@@ -98,7 +102,8 @@ export const createUpgradedEmbedBlock = (
 	// WordPress blocks can work on multiple sites, and so don't have patterns,
 	// so if we're in a WordPress block, assume the user has chosen it for a WordPress URL.
 	const isCurrentBlockWP =
-		providerNameSlug === WP_VARIATION.attributes.providerNameSlug ||
+		( WP_VARIATION &&
```
Comment (ntsekouras) [inline packages/block-library/src/embed/util.js:105]: If `WP_VARIATION` doesn't exist but `type=== WP_EMBED_TYPE` would result in true even if the wanted check would be to mark it `falsy` as we have unregistered the `WP` variation. This is the case you want to handle, no?

```
WP_VARIATION &&
( providerNameSlug === WP_VARIATION.attributes.providerNameSlug ||
	type === WP_EMBED_TYPE );
```

Comment (ockham) [reply packages/block-library/src/embed/util.js:105]: Ah yeah, good point, thank you!

**Comment by ockham** [comment]
Should be ready for review :slightly_smiling_face:

File: packages/block-library/src/embed/util.js [hunk 2]
```diff
@@ -121,7 +132,7 @@ export const createUpgradedEmbedBlock = (
 	// This is not the WordPress embed block so transform it into one.
 	return createBlock( DEFAULT_EMBED_BLOCK, {
 		url,
-		...WP_VARIATION.attributes,
+		...WP_VARIATION?.attributes,
```
Comment (mcsf) [inline packages/block-library/src/embed/util.js:135]: Even if `?.` prevents type errors, is this still correct when there is no WP_VARIATION?

**Review by mcsf — COMMENTED** [review]
Looks good in principle, but I haven't tested. I'll defer to @ntsekouras for review and approval. :)

File: packages/block-library/src/embed/util.js [hunk 3]
```diff
@@ -95,11 +101,16 @@ export const createUpgradedEmbedBlock = (
...
+	const WP_VARIATION = getBlockVariations( DEFAULT_EMBED_BLOCK )?.find(
+		( { name } ) => name === 'wordpress'
+	);
+
 	// WordPress blocks can work on multiple sites, and so don't have patterns,
 	// so if we're in a WordPress block, assume the user has chosen it for a WordPress URL.
 	const isCurrentBlockWP =
-		providerNameSlug === WP_VARIATION.attributes.providerNameSlug ||
-		type === WP_EMBED_TYPE;
+		WP_VARIATION && // Is the WordPress variation of the embed block registered?
```
Comment (ntsekouras) [inline packages/block-library/src/embed/util.js:111]: This comment/question feels like a personal note. Can you rephrase it to something like: `Check if .... is registered` ?

File: packages/block-library/src/embed/util.js (hunk 2 above)
Comment (ntsekouras) [reply packages/block-library/src/embed/util.js:135]: @mcsf is right. In the above check in line 129

```
if ( ! preview || ! isFromWordPress( preview.html ) || isCurrentBlockWP ) {
		return;
}
```
should also check if  `WP_VARIATION` exists and if not it should return nothing so as to not convert to WP block.

```
if ( ! preview  || ! WP_VARIATION  || ! isFromWordPress( preview.html ) || isCurrentBlockWP) {
		return;
}
```

The reason for not suggesting to remove `WP_VARIATION` from `isCurrentBlockWP` variable is that in that check the `providerNameSlug` can be `undefined` in some cases ( called from image block for example ) and that could lead to false positives for a WP block.

File: packages/block-library/src/embed/test/index.js [hunk 4]
```diff
@@ -75,13 +93,21 @@ describe( 'utils', () => {
 	} );
 	describe( 'createUpgradedEmbedBlock', () => {
 		describe( 'do not create new block', () => {
-			it( 'when block type does not exist', () => {
+			it( 'when block variation does not exist', () => {
```
Comment (ntsekouras) [inline packages/block-library/src/embed/test/index.js:96]: The previous test should not be removed ( `it( 'when block type does not exist', () => {...` ).

It tests the case of `core/embed` block being registered, in utils line 100.
```
if ( ! url || ! getBlockType( DEFAULT_EMBED_BLOCK ) ) return;
```

**Review by ntsekouras — COMMENTED** [review]
This looks good 👍  - I left some small comments though.

File: packages/block-library/src/embed/util.js (hunk 3 above)
Comment (ockham) [reply packages/block-library/src/embed/util.js:111]: Can do! I tend to leave out the `Check if ... is` part when commenting conditionals (since it takes up space), but I don't mind adding it back.

File: packages/block-library/src/embed/test/index.js (hunk 4 above)
Comment (ockham) [reply packages/block-library/src/embed/test/index.js:115]: d80d2c38b8113a8b11fa698a8b65453213f192ad

File: packages/block-library/src/embed/util.js (hunk 2 above)
Comment (ockham) [reply packages/block-library/src/embed/util.js:135]: The conditionals in `createUpgradedEmbedBlock` seem a bit convoluted and hard to read :confused: It seems like if we need `WP_VARIATION` in `isCurrentBlockWP`, but later on need to check for both criteria again, there's some room for improvement. Maybe we can shuffle some of these checks around to move early, and evaluate some stuff only if we have ruled out a few cases :thinking:

Comment (ockham) [reply packages/block-library/src/embed/util.js]: 2f81a0ff7f

**Comment by ockham** [comment]
Thanks for the feedback! Should be ready for another look :slightly_smiling_face:

File: packages/block-library/src/embed/util.js [hunk 5]
```diff
@@ -114,14 +119,24 @@ export const createUpgradedEmbedBlock = (
 		} );
 	}
 
+	const wpVariation = getBlockVariations( DEFAULT_EMBED_BLOCK )?.find(
+		( { name } ) => name === 'wordpress'
```
Comment (ntsekouras) [inline packages/block-library/src/embed/util.js:123]: Some comment as above to retrieve hardcoded value. Since `name` and `providerNameSlug` are the same now, choose what fits more. For example if you choose only the providerName, change this `find` check to use the same property.

**Review by ntsekouras — APPROVED** [review]
I left a comment to change the hardcoded value `wordpress` but when this change happens let's land this! 👍 

Thanks @ockham!

File: packages/block-library/src/embed/util.js (hunk 5 above)
Comment (ockham) [reply packages/block-library/src/embed/util.js:123]: TBH for the sake of future refactoring, I'd rather stick with `name` here, since it'll be one thing less to think about for future devs who'd like to phase out `providerNameSlug`. (Semantically, it also makes more sense to say that we're looking for the right block variation by its name, rather than its provider name.)

File: packages/block-library/src/embed/util.js [hunk 6]
```diff
@@ -98,8 +104,7 @@ export const createUpgradedEmbedBlock = (
 	// WordPress blocks can work on multiple sites, and so don't have patterns,
 	// so if we're in a WordPress block, assume the user has chosen it for a WordPress URL.
 	const isCurrentBlockWP =
-		providerNameSlug === WP_VARIATION.attributes.providerNameSlug ||
-		type === WP_EMBED_TYPE;
+		providerNameSlug === 'wordpress' || type === WP_EMBED_TYPE;
```
[the remark this answers is not in the document]
Comment (ockham) [reply packages/block-library/src/embed/util.js:107]: TBH given my [other comment](https://github.com/WordPress/gutenberg/pull/24559#discussion_r474144460) plus the fact that these are currently two different fields, using a constant to compare them both to has the potential to make this more confusing IMO. Using the same constant kind of implies that both fields are somehow functionally equivalent, which they aren't. Using a string literal in both cases (which just happens to be identical) carries that notion much less.

(To clarify, my previous suggestion to drop `providerNameSlug` was solely meant for the variations definition JSON file -- I didn't mean to remove the provider name check altogether for embeds that we're fetching from a given URL -- at least not before I understand it better :sweat_smile: )

**Comment by ockham** [comment]
> I left a comment to change the hardcoded value `wordpress` but when this change happens let's land this! 
> 
> Thanks @ockham!

Thanks @ntsekouras! I replied to both comments, essentially suggesting to keep the `wordpress` literals around :sweat_smile: Curious to hear your thoughts.

**Comment by ntsekouras** [comment]
> Using the same constant kind of implies that both fields are somehow functionally equivalent, which they aren't. Using a string literal in both cases (which just happens to be identical) carries that notion much less.

>essentially suggesting to keep the wordpress literals 

I agree with that but still the `name` don't you think it should be retrieved from `variations.js` and not be hardcoded? Even with the removal of the `providerSlugName`, the name would still be there. Maybe one and one 😅 ? What do you think?

Either way this is just a small comment and that's why I had pre-approved :)

**Comment by ockham** [comment]
> > Using the same constant kind of implies that both fields are somehow functionally equivalent, which they aren't. Using a string literal in both cases (which just happens to be identical) carries that notion much less.
> 
> > essentially suggesting to keep the wordpress literals
> 
> I agree with that but still the `name` don't you think it should be retrieved from `variations.js` and not be hardcoded? Even with the removal of the `providerSlugName`, the name would still be there. Maybe one and one :sweat_smile: ? What do you think?

Not sure I'm totally following :sweat_smile: -- we use the `name` to [look up](https://github.com/WordPress/gutenberg/pull/24559/files#diff-56fcc4a2c51b3f80f9ae35c2b3b94b34R122-R123) the `wordpress` variation object from the array in `variations.js`. We kind of needed _something_ hard-coded for the lookup, don't we? :sweat_smile:

**Comment by ntsekouras** [comment]
You're right. Got a bit stuck there. Land this 💯
